lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+};
+
+struct ExtractInstancesPass
+    : public ExtractInstancesBase<ExtractInstancesPass> {
+  void runOnOperation() override;
+  void collectAnnos();
+  void collectAnno(InstanceOp inst, Annotation anno);
+  void extractInstances();
+
+private:
+  bool anythingChanged;
+  bool anyFailures;
+
+  InstanceGraph *instanceGraph = nullptr;
+
+  /// The target directory to output instances into. Can be changed through
+  /// `firrtl.transforms.BlackBoxTargetDirAnno` annotations.
+  StringRef targetDir;
+
+  /// The modules in the design that are annotated with one or more annotations
+  /// relevant for instance extraction.
+  DenseMap<Operation *, SmallVector<Annotation, 1>> annotatedModules;
+
+  /// All modules that are marked as DUT themselves, or that have a DUT parent
+  /// module.
+  DenseSet<Operation *> dutModules;
+
+  /// A worklist of instances that need to be moved.
+  SmallVector<std::pair<InstanceOp, ExtractionInfo>> extractionWorklist;
+};
+} // end anonymous namespace
+
+/// Emit the annotated source code for black boxes in a circuit.
+void ExtractInstancesPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+  anythingChanged = false;
+  anyFailures = false;
+  targetDir = ".";
+  annotatedModules.clear();
+  dutModules.clear();
+  extractionWorklist.clear();
+
+  // Determine the target directory from the annotations on the circuit
+  // operation.
+  for (auto anno : AnnotationSet(circuitOp)) {
+    // Handle target dir annotation.
+    if (anno.isClass(blackBoxTargetDirAnnoClass)) {
+      if (auto target = anno.getMember<StringAttr>("targetDir")) {
+        targetDir = target.getValue();
+        continue;
+      }
+      circuitOp->emitError(blackBoxTargetDirAnnoClass)
+          << " annotation missing \"targetDir\" attribute";
+      anyFailures = true;
+      continue;
+    }
+  }
+  if (anyFailures) {
+    signalPassFailure();
+    return;
+  }
+
+  // Walk the IR and gather all the annotations relevant for extraction that
+  // appear on instances and the instantiated modules.
+  instanceGraph = &getAnalysis<InstanceGraph>();
+  collectAnnos();
+  if (anyFailures) {
+    signalPassFailure();
+    return;
+  }
+
+  // Actually move instances upwards.
+  extractInstances();
+
+  // TODO: Group instances if so requested.
+
+  // If nothing has changed we can preseve the analysis.
+  LLVM_DEBUG(llvm::dbgs() << "\n");
+  if (!anythingChanged)
+    markAllAnalysesPreserved();
+  markAnalysesPreserved<InstanceGraph>();
+}
+
+static bool isAnnoInteresting(Annotation anno) {
+  return anno.isClass(extractBlackBoxAnnoClass);
+}
+
+/// Gather the modules and instances annotated to be moved by this pass. This
+/// populates the corresponding lists and maps of the pass.
+void ExtractInstancesPass::collectAnnos() {
+  CircuitOp circuit = getOperation();
+
+  // Gather the annotations on modules. These complement the later per-instance
+  // annotations.
+  circuit.walk([&](FModuleLike module) {
```

Comment:
Unrelated comment: we should really just add an optional `DUT` member to the `circuitOp`. The "search all the modules for maybe a DUT" is repeated a lot of places. 🥲

Reply:
Nice! Switched that to `getOps` which is nicer and doesn't need that weird `WalkResult::skip()` at the end.

Yeah. On the one hand the DUT handling feels so dirty that I'm almost glad we're not codifying it in the `CircuitOp` -- but it would make a few things easier.